### PR TITLE
in-toto-keygen: a command line utility for generating keys

### PR DESCRIPTION
**Pull request by isachit**
A CLI tool for creating key files, and dumping them with filename , filename.pub for private and public keys respectively.
General Usage:
`python in_toto_keygen.py [-p] <filename>`

**Issue Referenced** - #96 

Please note the progress:
- [x] Write the script
- [x] Add Exceptions
- [x] Add Documentation
- [x] Write the unit tests and run them
- [ ] Fix any other issues, and add other functionalities.

@lukpueh and @SantiagoTorres.

UPDATE 1: Added @lukpueh in the author field since he is the one who actually wrote the two functions I have used.
UPDATE 2: It's pretty much in place. Just removing the unnecessary history remains.

### CODE REVIEW

**Comment by isachit** [comment]
I wonder, why aren't @SantiagoTorres  and @lukpueh  in the participants section. :| :/ And also, The travis CI build, along with coveralls. Hm.. Fishy.

**Comment by isachit** [comment]
Oh yeeesss!!!! Finally!! :D

File: test/test_in_toto_keygen.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+  def test_prompt_password(self):
+      """Call password prompt. """
+      password = "123456"
+      with patch("getpass.getpass", return_value=password):
+          self.assertEqual(prompt_password(), password)
+
+
+if __name__ == '__main__':
+  unittest.main(buffer=True)
```
Comment (SantiagoTorres) [inline test/test_in_toto_keygen.py]: Can you end this line with a newline? This usually confuses linecount algorithms and causes off-by-ones on diffing

File: test/test_in_toto_keygen.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+"""
```
Comment (SantiagoTorres) [inline test/test_in_toto_keygen.py:1]: In addition to the existing tests, I'm wondering if it'd be nice to have:

1. Create a keypair, and load the public and private keys
2. Create a keypair with different length parameters and load it

**Review by SantiagoTorres — COMMENTED** [review]
Looking good! 👍

File: test/test_in_toto_keygen.py (hunk 2 above)
Comment (isachit) [reply test/test_in_toto_keygen.py:1]: Alright. On it. :)

File: test/test_in_toto_keygen.py (hunk 1 above)
Comment (isachit) [reply test/test_in_toto_keygen.py]: Okay!

**Comment by isachit** [comment]
@SantiagoTorres  I guess, this is what you suggested right? or?

File: in_toto/in_toto_keygen.py [hunk 3]
```diff
@@ -14,7 +14,9 @@
 
 <Purpose>
   A CLI tool for creating key files, and dumping them with <filename>,
-  <filename>.pub for private and public keys respectively.
+  <filename>.pub for private and public keys respectively. It also takes an
+  integer as an input, which specifies the length of the RSA key to be
+  generated. By default it is set as 3072.
```
Comment (SantiagoTorres) [inline in_toto/in_toto_keygen.py]: Yeah this is good.

**Review by SantiagoTorres — COMMENTED** [review]
Looks good to me. @lukpueh what do you think?

File: in_toto/in_toto_keygen.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  try:
+    if args.prompt:
+      prompt_generate_and_write_rsa_keypair(args.name, args.bits)
+      sys.exit(0)
+    else:
+      generate_and_write_rsa_keypair(args.name, args.bits, password=None)
+      sys.exit(0)
+
+  except Exception as e:
+    print('The following error occurred', e)
```
Comment (lukpueh) [inline in_toto/in_toto_keygen.py]: Please use `log.error` instead of print.

File: in_toto/in_toto_keygen.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                            "private key with the same before storing")
+
+  in_toto_args.add_argument("name", type=str,
+                            help="The filename of the resulting key files",
+                            metavar="<filename>")
+
+  in_toto_args.add_argument("bits", nargs= "?", default=3072, type=int,
+                            help="The key size, or key length, of the RSA "
+                            "key.  'bits' must be 2048, or greater, and a "
+                            "multiple of 256.", metavar="<bits>")
```
Comment (lukpueh) [inline in_toto/in_toto_keygen.py]: Latter is only [partially true](https://github.com/secure-systems-lab/securesystemslib/blob/master/securesystemslib/keys.py#L211-L214) also see the [default crypto library](https://github.com/secure-systems-lab/securesystemslib/blob/master/securesystemslib/settings.py#L49).
I think you can just drop the info here.

**Review by lukpueh — CHANGES_REQUESTED** [review]
Great work!

 I wonder though why you copied over the code from `in_toto.util`. I thought we had agreed that you will try to use `securesystemslib.keys` directly, or if that's not possible, just use `in_toto.util`.

I'd rather not have the same code twice, especially since I'd like to move that particular code to `securesystemslib.keys` altogether. 

👍  for adding the key length parameter. 
I'd welcome if you changed `in_toto.util. generate_and_write_rsa_keypair` to accept an optional bits parameter and then use that function here.

**Comment by lukpueh** [comment]
Great work! Thanks for the PR, @isachit. Out of curiosity, why did you call the branch *nonuke*?

**Comment by isachit** [comment]
@SantiagoTorres I see what you did there! :D :P Haha. There were some issues with rebasing and removing the commit history not relevant to this PR. So @SantiagoTorres  helped me with that, and while doing it, he made the branch nonuke.

**Comment by SantiagoTorres** [comment]
[¯\\_(ツ)_/¯](http://i.imgur.com/7A2QTxW.gif)

**Pull request merged**